managed-ledger/src/main/java/org/apache/bookkeeper/mledger/util/RangeCache.java

```diff
@@ -106,7 +106,7 @@ K getKey() {
             return localKey;
         }
 
-        V getValue(K key) {
+        V getValue(K key, boolean requireSameKeyInstance) {
```

Comment:
what about adding a second method that is more lenient instead of increasing the complexity of the existing ?
this is called on the hot paths, I think that it is worth to add some code duplication

Reply:
Actually, the code becomes more complex when adding more methods. I've added javadoc to explain parameter value. I'll push the changes soon.

Reply:
@eolivelli I made a second attempt to refactor the code based on your review comment. PTAL. The RangeCache design is intentionally racy without any exclusive locks and since the values are reference counted recycled objects, it has resulted in this solution where consistency is checked with the solution that is fairly complex. 
After applying this solution, I haven't seen reports about bugs where the root cause was that there were races in RangeCache eviction and cache usage. Some of the complexity would be reduced if we wouldn't recycle the EntryWrapper instances in RangeCache. However that comes with a tradeoff regarding GC. Extensively testing performance could shade more light in actual differences in practice.